fix: allow doesnotcontains filters on hyperparameter column

ashtonG opened the pull request:
## Description

this fixes an issue where the sql generated by a filter containing a doesNotContain clause on a hyperparameter column was invalid

## Test Plan

in the experiment list, add a filter for a hyperparameter column (starts with "hp."). set the condition to "does not contain", and ensure the value isn't blank. The filter should correctly filter the experiment list.



## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
WEB-1988


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

-->

NicholasBlaskey commented on master/internal/experiment_filter.go:
```diff
@@ -214,7 +214,7 @@ func hpToSQL(c string, filterColumnType *string, filterValue *interface{},
 			queryString = fmt.Sprintf(`(CASE
 				WHEN config->'hyperparameters'->%s->>'type' = 'const' THEN config->'hyperparameters'->%s->>'val' NOT LIKE %s
-				WHEN config->'hyperparameters'->%s->>'type' = 'categorical' THEN (config->'hyperparameters'->%s->>'vals')::jsonb %s %s) IS NOT TRUE
+				WHEN config->'hyperparameters'->%s->>'type' = 'categorical' THEN (config->'hyperparameters'->%s->>'vals')::jsonb %s %s IS NOT TRUE
```
drive by review since this came up in my slack threads

Can we include tests? This endpoint is really hard to change just because of all the possible dynamic sql that can build. It would be cool if it was exhaustively tested / close to 100% test coverage.

ashtonG replied:
i would rather that be in a separate pr -- as you mentioned, there's a lot of possible dynamic sql that can be built. exhaustive testing would add a lot more code to review.

NicholasBlaskey replied:
Can we include a test for just this regression then?

and make a separate ticket

ashtonG replied:
sounds like a good compromise -- will do!

NicholasBlaskey replied:
I guess this bug is the reason why the current tests are not great. They didn't catch this bug.

Like it's great that we can assert the sql looks some way but it doesn't catch generating wrong sql. Nothing in that test makes me think this code is actually fixing the issue.

What I think would be a lot better test would be to actually run a query with Not Contains against the database and make sure that it doesn't error and filters as expected.

ashtonG replied:
gotcha -- added!

NicholasBlaskey commented on master/internal/experiment_filter_intg_test.go:
```diff
@@ -0,0 +1,96 @@
+	pid := projResp.Project.Id
+
+	// nolint: exhaustruct
+	activeConfig0 := schemas.WithDefaults(schemas.Merge(minExpConfig, expconf.ExperimentConfig{
+		RawHyperparameters: expconf.Hyperparameters{
+			"hyperparameter": expconf.HyperparameterV0{
+				RawConstHyperparameter: &expconf.ConstHyperparameterV0{
+					RawVal: "foo",
+				},
+			},
+		},
+	}))
+	job0ID := uuid.New().String()
+
+	startTime := time.Unix(123123123, int64(1329012309*time.Nanosecond))
+	endTime := time.Unix(423123123, int64(999813239*time.Nanosecond))
+	require.WithinDuration(t,
+		endTime, timestamppb.New(endTime).AsTime(), time.Millisecond)
+
+	exp0 := &model.Experiment{
+		StartTime: startTime,
+		EndTime:   &endTime,
+		JobID:     model.JobID(job0ID),
+		Archived:  false,
+		State:     model.PausedState,
+		Notes:     "notes",
+		Config:    activeConfig0.AsLegacy(),
+		OwnerID:   ptrs.Ptr(model.UserID(1)),
+		ProjectID: int(pid),
+	}
+	require.NoError(t, api.m.db.AddExperiment(exp0, []byte{1, 2, 3}, activeConfig0))
```
should be able to use something like

```
paramNames := []string{"foo"}
db.RequireMockExperimentParams(t, api.m.db, user, MockExperimentParams{
    HParamNames: &paramNames,
}, projectID)
```

NicholasBlaskey commented on master/internal/experiment_filter_intg_test.go:
```diff
@@ -0,0 +1,96 @@
+	query := db.Bun().NewSelect().
+		TableExpr("experiments as e")
+
+	_, queryErr := filter.toSQL(query)
```
can we just call the API instead?

NicholasBlaskey reviewed (approved):
LGTM assuming the tests pass

 thanks for writing the tests

ashtonG commented:
hey @NicholasBlaskey -- finally got time to circle around to your review. i've implemented your suggestions and moved the test to the experiment integration test suite -- does this look okay?
